Address safer CPP failures in platform/network/cocoa

cdumez opened the pull request:
#### 8b6bee932bfd57c53cbe234e87cb820d8328cff5
<pre>
Address safer CPP failures in platform/network/cocoa
<a href="https://bugs.webkit.org/show_bug.cgi?id=290642">https://bugs.webkit.org/show_bug.cgi?id=290642</a>

Reviewed by Timothy Hatcher, Geoffrey Garen, and Darin Adler.

* Source/WTF/wtf/RetainPtr.h:
(WTF::lazyInitialize):
* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebCore/platform/network/cocoa/CookieCocoa.mm:
(WebCore::portVectorFromList):
(WebCore::portStringFromVector):
(WebCore::cookieCreated):
(WebCore::cookieExpiry):
(WebCore::Cookie::operator NSHTTPCookie * _Nullable  const):
* Source/WebCore/platform/network/cocoa/CookieStorageObserver.mm:
(WebCore::CookieStorageObserver::startObserving):
* Source/WebCore/platform/network/cocoa/CredentialCocoa.h:
(WebCore::Credential::encodingRequiresPlatformData const):
* Source/WebCore/platform/network/cocoa/CredentialCocoa.mm:
(WebCore::Credential::Credential):
* Source/WebCore/platform/network/cocoa/NetworkLoadMetrics.mm:
(WebCore::copyTimingData):
* Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
(WebCore::NetworkStorageSession::setAllCookiesToSameSiteStrict):
(WebCore::policyProperties):
(WebCore::NetworkStorageSession::capExpiryOfPersistentCookie):
(WebCore::NetworkStorageSession::cookiesForSessionAsVector const):
(WebCore::adjustScriptWrittenCookie):
(WebCore::parseDOMCookie):
(WebCore::NetworkStorageSession::setCookiesFromDOM const):
(WebCore::NetworkStorageSession::deleteCookie const):
(WebCore::NetworkStorageSession::getHostnamesWithCookies):
(WebCore::NetworkStorageSession::deleteAllCookies):
(WebCore::NetworkStorageSession::domCookiesForHost):
* Source/WebCore/platform/network/cocoa/ProtectionSpaceCocoa.mm:
(WebCore::type):
(WebCore::scheme):
(WebCore::ProtectionSpace::ProtectionSpace):
(WebCore::ProtectionSpace::nsSpace const):
* Source/WebCore/platform/network/cocoa/RangeResponseGenerator.mm:
(WebCore::RangeResponseGenerator::Data::shutdownResource):
(WebCore::RangeResponseGenerator::removeTask):
(WebCore::RangeResponseGenerator::willSynthesizeRangeResponses):
* Source/WebCore/platform/network/cocoa/ResourceRequestCocoa.mm:
(WebCore::ResourceRequest::doUpdateResourceRequest):
(WebCore::siteForCookies):
(WebCore::ResourceRequest::doUpdatePlatformRequest):
(WebCore::ResourceRequest::doUpdatePlatformHTTPBody):
* Source/WebCore/platform/network/cocoa/ResourceResponseCocoa.mm:
(WebCore::ResourceResponse::initNSURLResponse const):
(WebCore::ResourceResponse::platformCertificateInfo const):
* Source/WebCore/platform/network/cocoa/WebCoreNSURLSession.mm:
(-[WebCoreNSURLSession finishTasksAndInvalidate]):
(-[WebCoreNSURLSessionDataTask initWithSession:identifier:request:targetDispatcher:]):
(-[WebCoreNSURLSessionDataTask resource:receivedResponse:completionHandler:]):
(-[WebCoreNSURLSessionDataTask resource:receivedData:]):
(-[WebCoreNSURLSessionDataTask resource:receivedRedirect:request:completionHandler:]):
(-[WebCoreNSURLSessionDataTask _resource:loadFinishedWithError:metrics:]):
* Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.mm:
(WebKit::cookiesByCappingExpiry):

Canonical link: <a href="https://commits.webkit.org/292902@main">https://commits.webkit.org/292902@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f7b1354d6026837a381cb3bb635a597b6e9ea04a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97390 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17015 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7231 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102477 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47918 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/99435 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17308 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25465 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74195 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31381 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100393 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13098 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88073 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54538 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12862 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5958 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47361 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/90066 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82894 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6039 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104497 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/96012 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24469 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/17839 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/83241 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/24841 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84202 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82661 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27209 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/4896 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/18017 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15734 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24431 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/29599 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/119638 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24253 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/33584 "Found 4 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/es6/ES6MathAPIs.js.default, microbenchmarks/let-const-tdz-environment-parsing-and-hash-consing-speed.js.no-llint, wasm.yaml/wasm/stress/spilled-constant-block-argument.js.wasm-bbq-no-consts, wasm.yaml/wasm/stress/spilled-constant-block-result.js.wasm-bbq-no-consts (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27567 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25827 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/platform/network/cocoa/CookieCocoa.mm:
```diff
@@ -36,17 +36,16 @@
     return Vector<uint16_t>(portList.count, [portList](size_t i) {
-        NSNumber *port = portList[i];
-        return port.unsignedShortValue;
+        return portList[i].unsignedShortValue;
```
Changes like this make me wonder about the static analyzer’s quality. The new code is exactly as safe as the old code, but somehow does not yield the same warning. If we need a `RetainPtr`, whether it’s an intermediate value in an expression or a local variable doesn’t matter.

cdumez replied:
cc @rniwa to comment. I do think that it shouldn't have required a `RetainPtr`.

rniwa replied:
Hm... I'm a bit surprised that the static analyzer doesn't generate a warning after this change. I would have expected it to generate a call argument warning since any Obj-C property access like `unsignedShortValue` is a selector invocation.

rniwa replied:
ok, I think we'd get a call argument warning after this change like this:
```
CookieCocoa.mm:39:16: warning: Reciever is unretained and unsafe [alpha.webkit.UnretainedCallArgsChecker]
   39 |         return portList[i].unsignedShortValue;
      |                ^~~~~~~~~~~
```

This is expected because `unsignedShortValue` is an Objective-C message send and can have side effects.

cdumez replied:
Maybe it will generate a warning. I haven't un-skipped files yet because I was waiting for Brianna to fix the bot.

rniwa replied:
So there are a couple of options here. We could deploy `RetainPtr` here out of principle. Alternatively, we could hard code a list of properties for `NSNumber` and other basic NS classes and allow them to be invoked without a `RetainPtr`.

rniwa replied:
Yet another option is to manually suppress this warning for this line because we know at least the common implementation of `unsignedShortValue` doesn't have a side effect.

It's a bit of risky business though because ObjC selectors can be overridden / fiddled with for a specific object. It's possible that someone could create a new subclass of `NSNumber` which implements `unsignedShortValue` with a serious side effect.

cdumez replied:
I will retain.

darinadler commented on Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
```diff
@@ -267,9 +267,9 @@ - (CFURLRef)_cfurl;
 static RetainPtr<NSDictionary> policyProperties(const SameSiteInfo& sameSiteInfo, NSURL *url)
 {
-    static NSURL *emptyURL = [[NSURL alloc] initWithString:@""];
+    static NeverDestroyed<RetainPtr<NSURL>> emptyURL = adoptNS([[NSURL alloc] initWithString:@""]);
```
I wonder if we should have a single shared one of these returned by a helper function rather than a local one here. Should search for other `initWithString:@""`.

cdumez replied:
Interesting. There is also `URL::emptyCFURL()` which constructs a NSURL internally. However, it doesn't use a static variable.

darinadler commented on Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
```diff
@@ -420,15 +420,15 @@ - (CFURLRef)_cfurl;
 
     Vector<Cookie> cookiesVector;
-    NSString *name = options.name;
+    RetainPtr<NSString> name = static_cast<NSString *>(options.name);
```
```suggestion
    RetainPtr name = options.name.protectedNSString();
```
While it’s not great to call a “protected” function when also putting something into a RetainPtr, I think it’s worse to write a cast rather than a named function. The reason we have the `operator NSString *` is so we can write things in a terse way, but a named function is clearer when we have to invoke it explicitly.

cdumez replied:
How about I add a `nsString()` getter on String and call it? I'd prefer not to set an example initializing a local variable with a static function (though I hesitated when I wrote this PR).

darinadler replied:
Sure, I would name it `createNSString()` just like `createCFString()`. But I’m not sure what you mean, what is the static function?

cdumez replied:
Sorry, I meant a `protected` function, not a `static` function.

cdumez replied:
I will use createNSString() and createNSURL()

darinadler commented on Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
```diff
@@ -512,7 +512,7 @@ - (CFURLRef)_cfurl;
         return;
 
-    NSURL *cookieURL = url;
+    RetainPtr<NSURL> cookieURL = static_cast<NSURL *>(url);
```
Here the same “write a cast” thing comes up but we don’t have a function like `protectedNSString`. Maybe we should add a `createNSURL` function that returns a `RetainPtr`; nice to avoid the autorelease too. Also makes me realize we could just rename `protectedNSString` to `createNSString`.

darinadler commented on Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
```diff
@@ -597,13 +597,13 @@ static NSHTTPCookieAcceptPolicy httpCookieAcceptPolicy(CFHTTPCookieStorageRef co
     RetainPtr<NSArray> cookies = httpCookiesForURL(cookieStorage.get(), firstParty, std::nullopt, url, ThirdPartyCookieBlockingDecision::None);
 
-    NSString *cookieNameString = cookieName;
+    RetainPtr<NSString> cookieNameString = static_cast<NSString *>(cookieName);
```
```suggestion
    RetainPtr cookingNameString = cookieName.protectedNSString();
```

geoffreygaren reviewed (approved):
Please update expectations.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -219,7 +219,7 @@
         path = path.substring(1);
 
-    auto *cocoaPath = static_cast<NSString *>(path);
+    RetainPtr cocoaPath = path.createNSString();
```
All of the `WebKit/*/Extensions` files are ARC and these are not leaks.

I would prefer to keep the casts as-is, or front load the `.get()`.
```suggestion
    auto *cocoaPath = path.createNSString().get();
```

cdumez replied:
I am not trying to fix leaks. Darin said in his review that he preferred we call a `createNSString()` function rather than using a cast. It also has the benefit of not using autorelease unnecessarily. That said, I can totally use `auto *cocoaPath = path.createNSString().get();` like you suggest since this code uses ARC.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -256,25 +256,25 @@
     }
 
-    auto *resourceURL = static_cast<NSURL *>(resourceFileURLForPath(path));
+    RetainPtr resourceURL = resourceFileURLForPath(path).createNSURL();
```
```suggestion
    auto *resourceURL = resourceFileURLForPath(path).createNSURL().get();
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
```diff
@@ -523,9 +523,9 @@ - (void)_didDeleteLocalStorage:(NSNotification *)notification
     WKPreferences *preferences = protectedConfiguration()->webViewConfiguration().preferences;
 
-    NSString *cocoaFeatureName = static_cast<NSString *>(featureName);
+    RetainPtr cocoaFeatureName = featureName.createNSString();
```
```suggestion
    auto *cocoaFeatureName = featureName.createNSString().get();
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -256,25 +256,25 @@
     if (!resultData) {
         if (suppressErrors == SuppressNotFoundErrors::No)
-            outError = createError(Error::ResourceNotFound, WEB_UI_FORMAT_CFSTRING("Unable to find \"%@\" in the extension’s resources.", "WKWebExtensionErrorResourceNotFound description with file name", (__bridge CFStringRef)cocoaPath), API::Error::create(fileReadError));
+            outError = createError(Error::ResourceNotFound, WEB_UI_FORMAT_CFSTRING("Unable to find \"%@\" in the extension’s resources.", "WKWebExtensionErrorResourceNotFound description with file name", (__bridge CFStringRef)cocoaPath.get()), API::Error::create(fileReadError));
```
```suggestion
            outError = createError(Error::ResourceNotFound, WEB_UI_FORMAT_CFSTRING("Unable to find \"%@\" in the extension’s resources.", "WKWebExtensionErrorResourceNotFound description with file name", bridge_cast(cocoaPath)), API::Error::create(fileReadError));
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -256,25 +256,25 @@
-    if (!validateResourceData(resourceURL, resultData, &validationError)) {
-        outError = createError(Error::InvalidResourceCodeSignature, WEB_UI_FORMAT_CFSTRING("Unable to validate \"%@\" with the extension’s code signature. It likely has been modified since the extension was built.", "WKWebExtensionErrorInvalidResourceCodeSignature description with file name", (__bridge CFStringRef)cocoaPath), API::Error::create(validationError));
+    if (!validateResourceData(resourceURL.get(), resultData, &validationError)) {
+        outError = createError(Error::InvalidResourceCodeSignature, WEB_UI_FORMAT_CFSTRING("Unable to validate \"%@\" with the extension’s code signature. It likely has been modified since the extension was built.", "WKWebExtensionErrorInvalidResourceCodeSignature description with file name", (__bridge CFStringRef)cocoaPath.get()), API::Error::create(validationError));
```
```suggestion
        outError = createError(Error::InvalidResourceCodeSignature, WEB_UI_FORMAT_CFSTRING("Unable to validate \"%@\" with the extension’s code signature. It likely has been modified since the extension was built.", "WKWebExtensionErrorInvalidResourceCodeSignature description with file name", bridge_cast(cocoaPath)), API::Error::create(validationError));
```

cdumez committed:
Address safer CPP failures in platform/network/cocoa
https://bugs.webkit.org/show_bug.cgi?id=290642

Reviewed by Timothy Hatcher, Geoffrey Garen, and Darin Adler.

* Source/WTF/wtf/RetainPtr.h:
(WTF::lazyInitialize):
* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebCore/platform/network/cocoa/CookieCocoa.mm:
(WebCore::portVectorFromList):
(WebCore::portStringFromVector):
(WebCore::cookieCreated):
(WebCore::cookieExpiry):
(WebCore::Cookie::operator NSHTTPCookie * _Nullable  const):
* Source/WebCore/platform/network/cocoa/CookieStorageObserver.mm:
(WebCore::CookieStorageObserver::startObserving):
* Source/WebCore/platform/network/cocoa/CredentialCocoa.h:
(WebCore::Credential::encodingRequiresPlatformData const):
* Source/WebCore/platform/network/cocoa/CredentialCocoa.mm:
(WebCore::Credential::Credential):
* Source/WebCore/platform/network/cocoa/NetworkLoadMetrics.mm:
(WebCore::copyTimingData):
* Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
(WebCore::NetworkStorageSession::setAllCookiesToSameSiteStrict):
(WebCore::policyProperties):
(WebCore::NetworkStorageSession::capExpiryOfPersistentCookie):
(WebCore::NetworkStorageSession::cookiesForSessionAsVector const):
(WebCore::adjustScriptWrittenCookie):
(WebCore::parseDOMCookie):
(WebCore::NetworkStorageSession::setCookiesFromDOM const):
(WebCore::NetworkStorageSession::deleteCookie const):
(WebCore::NetworkStorageSession::getHostnamesWithCookies):
(WebCore::NetworkStorageSession::deleteAllCookies):
(WebCore::NetworkStorageSession::domCookiesForHost):
* Source/WebCore/platform/network/cocoa/ProtectionSpaceCocoa.mm:
(WebCore::type):
(WebCore::scheme):
(WebCore::ProtectionSpace::ProtectionSpace):
(WebCore::ProtectionSpace::nsSpace const):
* Source/WebCore/platform/network/cocoa/RangeResponseGenerator.mm:
(WebCore::RangeResponseGenerator::Data::shutdownResource):
(WebCore::RangeResponseGenerator::removeTask):
(WebCore::RangeResponseGenerator::willSynthesizeRangeResponses):
* Source/WebCore/platform/network/cocoa/ResourceRequestCocoa.mm:
(WebCore::ResourceRequest::doUpdateResourceRequest):
(WebCore::siteForCookies):
(WebCore::ResourceRequest::doUpdatePlatformRequest):
(WebCore::ResourceRequest::doUpdatePlatformHTTPBody):
* Source/WebCore/platform/network/cocoa/ResourceResponseCocoa.mm:
(WebCore::ResourceResponse::initNSURLResponse const):
(WebCore::ResourceResponse::platformCertificateInfo const):
* Source/WebCore/platform/network/cocoa/WebCoreNSURLSession.mm:
(-[WebCoreNSURLSession finishTasksAndInvalidate]):
(-[WebCoreNSURLSessionDataTask initWithSession:identifier:request:targetDispatcher:]):
(-[WebCoreNSURLSessionDataTask resource:receivedResponse:completionHandler:]):
(-[WebCoreNSURLSessionDataTask resource:receivedData:]):
(-[WebCoreNSURLSessionDataTask resource:receivedRedirect:request:completionHandler:]):
(-[WebCoreNSURLSessionDataTask _resource:loadFinishedWithError:metrics:]):
* Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.mm:
(WebKit::cookiesByCappingExpiry):

Canonical link: https://commits.webkit.org/292902@main
